sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -721,8 +716,55 @@ case class GetExternalRowField(
           "cannot be null.");
       }
 
-      final ${ctx.javaType(dataType)} ${ev.value} = $getField;
+      final Object ${ev.value} = ${row.value}.get($index);
      """
     ev.copy(code = code, isNull = "false")
   }
 }
+
+/**
+ * Validates the actual data type of input expression at runtime.  If it doesn't match the
+ * expectation, throw an exception.
+ */
+case class ValidateExternalType(child: Expression, expected: DataType)
```

Comment:
The problem is we can't trust it.... When users call `createDataFrame(rows, schema)`, we should definitely validate the passed-in rows. I think performance doesn't matter too much here, as this only happens at the beginning of the data flow. One potential issue may be that, `Dataset.map` can return row and users will provide a schema we should trust. However, I don't think we should expose `RowEncoder` to users and `Dataset.map` should never return a row.